### DISCUSSION TITLE
External monitor gets assigned to second workspace on boot even though internal monitor is off

### DISCUSSION THREAD

**Discussion by 9las** [i3/i3#4939] [Q&A] [answered]
Hi, when I boot my laptop, the external monitor gets assigned to the second workspace even though the internal monitor is off. I have set up my computer to use autorandr to automatically switch monitors at boot. The relevant lines from my i3 config are:
```
# auto configure monitors and use postscript to set wallpaper and launch polybar
exec_always --no-startup-id $HOME/.config/i3/scripts/auto-configure_monitors.sh
```

Content of the script `auto-configure_monitors.sh`:
```
#!/usr/bin/env bash

# Try to change configuration and handle if it is already loaded
if autorandr --change 2>&1 >/dev/null | grep "Config already loaded" >/dev/null; then

# Execute postswitch
    bash $HOME/.config/autorandr/postswitch

fi
```

Content of the script `postswitch`:

```
#!/usr/bin/env bash

# set wallpaper
feh --bg-fill $HOME/Pictures/Wallpapers/i-HZT8VDH-O.jpg

# launch polybar
$HOME/.config/polybar/launch.sh
```

I tried to add the following line to `.xinitrc`: `xrandr --output DP-2 --off --output DP-3 --auto` before the line `exec i3`. That resulted in that the external monitor `DP-3` got assigned to the first workspace.

Thus, it seems like the reason that my external monitor gets assigned to the second workspace is that the workspaces gets assigned before the command `autorandr --change` gets executed in the script `auto-configure_monitors.sh`.

Is there a way to execute a script using the i3 config before workspaces gets assigned? Maybe there are other ways to solve my issue?

**Comment by Rom1deTroyes** [comment]
I don't know if this can help, but you can specify on which ouput you want your workspace to reside with :
```config
workspace 1 output DP-3 primary
```

And if you turn off one monitor, the workspace must move to the other one (I use it the opposite way : activate my hdmi display open a new workspace on this screen, and whenever I do a `xrandr --output HDMI-0 --off`, it moves all the remaining workspaces from this monitor to the internal screen...
So maybe the last thing to do is cutting the output of the internal screen when you close the lid ?

**Comment by 9las** [comment, marked as answer]
I solved the issue by rewriting the scripts such that `autorandr --change` now gets executed before starting i3.
I have the following in `.xinitrc`:
```
# Auto configure monitors
autorandr --change

# Start window manager
exec i3
```

In the postswitch of autorandr I am now checking if i3 is the active window manager:
```
#!/usr/bin/env bash

# Check that i3 is the active window manager
if [[ $(xprop -root -notype | sed -n 's/^_NET_WM_NAME = "\(.*\)"$/\1/p') == "i3" ]]; then

# set wallpaper
    i3-msg 'exec --no-startup-id feh --bg-fill $HOME/Pictures/Wallpapers/i-HZT8VDH-O.jpg'

# launch polybar
    i3-msg 'exec --no-startup-id $HOME/.config/polybar/launch.sh'
fi
```

Content of `launch.sh`:
```
#!/usr/bin/env bash

# Terminate already running bar instances
polybar-msg cmd quit

# Wait until the processes have been shut down
while pgrep -x polybar >/dev/null; do sleep 1; done

# Launch Polybar, using default config location ~/.config/polybar/config.ini
for monitor_name in $(polybar --list-monitors | cut -d":" -f1); do
    MONITOR=$monitor_name polybar example &
done
```

I also have the following lines in my i3 config:

```
# Set wallpaper
exec --no-startup-id feh --bg-fill $HOME/Pictures/Wallpapers/i-HZT8VDH-O.jpg

# Launch polybar
exec --no-startup-id $HOME/.config/polybar/launch.sh
```

So upon boot monitors are first auto configured with `autorandr --change`. This activates the postswitch if the detected monitor configuration is not already loaded. The postswitch will however not do anything since i3 is not the active window manager yet. i3 will then start up and the wallpaper will be set and polybar will get launched.

After i3 has been started, changing the monitor configuration will due to the postswitch result in that the wallpaper gets reset and that polybar gets relaunched.
